File: app/javascript/controllers/empaque_controller.js
import { Controller } from "@hotwired/stimulus"

// C21-01 · El escaneo al empacar.
//
// Yusef, mostrando la bodega en vivo mientras empacaban: *"ahí están empacando,
// mirá… y aquí es donde hace falta, **es el pip pip pip**"*.
//
// El operario mira la pistola, no la pantalla, así que **lo que decide es el
// sonido**. Por eso cada resultado dispara su evento y la vista los cablea a
// `audio#success` / `audio#error` / `audio#alert`.
//
// La pistola dispara Enter al terminar de leer — la misma mecánica de
// /etiquetar. Y va con el guard de carrera de siempre (`_seq`): si el operario
// escanea dos seguidos, la respuesta vieja no puede pintar encima de la nueva
// ni sonar por un paquete que ya no está en pantalla.
export default class extends Controller {
  static targets = ["codigo", "aviso", "filas", "caja", "tarjeta", "candado", "conteo", "destino", "listaAbiertas"]
  static values = { escanearUrl: String, omitirUrl: String, activa: Number }

  connect() {
    this._seq = 0
    if (this.hasCodigoTarget) this.codigoTarget.focus()
  }

  // ── C23-11 · Varias cajas abiertas ────────────────────────────────────────
  //
  // Yusef: *"poder **seleccionar las tres cajas**"* · *"ellos arman tres cajas
  // y empiezan a meter los paquetes en **cualquier** caja"*.
  //
  // Un paquete sigue yendo en UNA caja: lo que cambia es que ya no hay que
  // **ir a buscarla**. Elegir caja era un `link_to` que recargaba la pantalla,
  // y con eso el campo de escaneo perdía el foco — con la pistola en la otra
  // mano, un clic por cada cambio de caja. Acá solo se mueve a dónde apunta el
  // POST, y el foco vuelve al campo siempre.

  elegirCaja(e) {
    const bloque = e.currentTarget.closest("[data-caja-id]")
    // Una caja cerrada no recibe escaneos: tocarla la abre, que es lo que el
    // operario quiso decir al tocarla.
    if (bloque.dataset.abierta !== "true") return this.alternarCaja(e)

    this._activar(Number(bloque.dataset.cajaId))
    this.codigoTarget.focus()
  }

  alternarCaja(e) {
    const bloque = e.currentTarget.closest("[data-caja-id]")

    this._post(bloque.dataset.alternarUrl, {})
      .then((data) => {
        // «Mínimo uno»: el servidor es el que decide, y si dice que no, lo dice
        // con todas las letras en vez de dejar la tarjeta a medio apagar.
        if (!data.ok) return this._mostrar("noEncontrado", data.mensaje)

        this.cajaTargets.forEach((caja) => {
          caja.dataset.abierta = String(data.abiertas.includes(Number(caja.dataset.cajaId)))
        })
        this._pintar()
        this._activar(data.activa)
      })
      .finally(() => this.codigoTarget.focus())
  }

  _activar(cajaId) {
    this.activaValue = cajaId
    const bloque = this.cajaTargets.find((c) => Number(c.dataset.cajaId) === cajaId)
    if (!bloque) return

    this.escanearUrlValue = bloque.dataset.escanearUrl
    this.omitirUrlValue = bloque.dataset.omitirUrl
    this._pintar()
  }

  // Las clases se arman acá y no en el servidor porque nada recarga: si el
  // pintado viviera solo en el ERB, la tarjeta activa se quedaría marcando la
  // caja anterior hasta el próximo refresh.
  _pintar() {
    const ACTIVA = ["border-cec-gold", "bg-cec-gold/5", "ring-2", "ring-cec-gold/30"]
    const ABIERTA = ["border-gray-200", "dark:border-gray-700", "bg-white", "dark:bg-gray-800",
                     "hover:shadow-lg", "hover:border-cec-gold/50"]
    const CERRADA = ["border-dashed", "border-gray-300", "dark:border-gray-700",
                     "bg-gray-50", "dark:bg-gray-900", "opacity-60"]

    this.cajaTargets.forEach((caja) => {
      const tarjeta = caja.querySelector("[data-empaque-target='tarjeta']")
      const abierta = caja.dataset.abierta === "true"
      const activa = Number(caja.dataset.cajaId) === this.activaValue && abierta

      tarjeta.classList.remove(...ACTIVA, ...ABIERTA, ...CERRADA)
      tarjeta.classList.add(...(activa ? ACTIVA : abierta ? ABIERTA : CERRADA))
      tarjeta.setAttribute("aria-pressed", String(activa))

      // El candado: los dos iconos están en el DOM y se alterna cuál se ve.
      const candado = caja.querySelector("[data-empaque-target='candado']")
      candado.querySelector("[data-icono='abierta']").classList.toggle("hidden", !abierta)
      candado.querySelector("[data-icono='cerrada']").classList.toggle("hidden", abierta)

      if (activa && this.hasDestinoTarget) {
        this.destinoTarget.textContent = tarjeta.querySelector("span").textContent.trim()
      }
    })

    this._pintarTabla()
  }

  // Lo que el servidor pintó una vez y acá ya no se recarga nunca: el título
  // que enumera las abiertas y las filas de las cajas que se cerraron. Sin
  // esto los dos se quedan diciendo el set del primer render — el título
  // listando cajas que ya nadie está llenando, y la tabla mostrando lo que hay
  // adentro de una caja cerrada.
  _pintarTabla() {
    const abiertas = this.cajaTargets.filter((c) => c.dataset.abierta === "true")

    if (this.hasListaAbiertasTarget) {
      this.listaAbiertasTarget.textContent = abiertas
        .map((c) => c.querySelector("[data-empaque-target='tarjeta'] span").textContent.trim())
        .join(", ")
    }

    if (!this.hasFilasTarget) return
    const ids = abiertas.map((c) => c.dataset.cajaId)
    this.filasTarget.querySelectorAll("tr[data-caja-id]").forEach((fila) => {
      fila.classList.toggle("hidden", !ids.includes(fila.dataset.cajaId))
    })
  }

  teclado(e) {
    if (e.key !== "Enter") return
    e.preventDefault()
    this.escanear()
  }

  escanear() {
    const codigo = this.codigoTarget.value.trim()
    if (codigo === "") return

    const consulta = (this._seq += 1)
    this.codigoTarget.value = ""

    this._post(this.escanearUrlValue, { codigo })
      .then((data) => {
        if (consulta !== this._seq) return  // llegó tarde: habla de otro escaneo
        this._resolver(data)
      })
      .catch(() => {
        this.dispatch("noEncontrado")
        this._mostrar("noEncontrado", "No se pudo consultar. Probá de nuevo.")
      })
  }

  // C21-01 · «Omitir»: lo mete igual aunque el tipo no concuerde. La Fase 12 lo
  // pidió *"para no trabar la operación cuando algo no cuadra"*.
  omitir(e) {
    const paqueteId = e.currentTarget.dataset.paqueteId
    this._post(this.omitirUrlValue, { paquete_id: paqueteId })
      .then((data) => this._resolver(data))
      .finally(() => this.codigoTarget.focus())
  }

  // Los `dispatch` van con el nombre **literal**, uno por rama. Con un
  // `dispatch(variable)` el sonido igual sonaría, pero `sonidos_cableados_test`
  // no podría probarlo leyendo el archivo — y ése es justo el bug que ese lint
  // existe para atrapar: un cable suelto en Stimulus no tira error, no ensucia
  // la consola; simplemente no suena.
  _resolver(data) {
    switch (data.resultado) {
      case "ok":            this.dispatch("ok"); break
      case "tipo_distinto": this.dispatch("tipoDistinto"); break
      case "ya_empacado":   this.dispatch("yaEmpacado"); break
      default:              this.dispatch("noEncontrado")
    }
    this._mostrar(this._tono(data.resultado), data.mensaje, data.paquete_id)
    if (data.fila) {
      this._agregarFila(data.fila)
      this._sumarAlConteo()
    }
    this.codigoTarget.focus()
  }

  _tono(resultado) {
    return { ok: "ok", tipo_distinto: "tipoDistinto", ya_empacado: "yaEmpacado" }[resultado] || "noEncontrado"
  }

  _mostrar(evento, mensaje, paqueteId) {
    if (!this.hasAvisoTarget) return

    const tonos = {
      ok: "bg-cec-teal/10 text-cec-teal-dark",
      tipoDistinto: "bg-red-50 text-red-800",
      yaEmpacado: "bg-cec-gold/15 text-cec-navy",
      noEncontrado: "bg-cec-gold/15 text-cec-navy"
    }
    this.avisoTarget.className = `mt-4 rounded-lg p-3 text-sm ${tonos[evento] || tonos.noEncontrado}`
    this.avisoTarget.textContent = mensaje
    this.avisoTarget.classList.remove("hidden")

    if (evento === "tipoDistinto" && paqueteId) this._botonOmitir(paqueteId)
  }

  _botonOmitir(paqueteId) {
    const boton = document.createElement("button")
    boton.type = "button"
    boton.textContent = "Meterlo igual (omitir)"
    boton.className = "ml-3 underline font-medium"
    boton.dataset.paqueteId = paqueteId
    boton.dataset.action = "click->empaque#omitir"
    this.avisoTarget.appendChild(boton)
  }

  // El «N pqt» de la tarjeta activa. Sin esto la cuenta se queda en lo que
  // había al cargar la pantalla, y como acá ya no se recarga nunca, se quedaría
  // vieja todo el turno.
  _sumarAlConteo() {
    const bloque = this.cajaTargets.find((c) => Number(c.dataset.cajaId) === this.activaValue)
    const conteo = bloque?.querySelector("[data-empaque-target='conteo']")
    if (!conteo) return

    const n = parseInt(conteo.textContent, 10)
    conteo.textContent = `${(Number.isInteger(n) ? n : 0) + 1} pqt`
  }

  _agregarFila(fila) {
    if (!this.hasFilasTarget) return
    const tr = document.createElement("tr")
    tr.dataset.cajaId = String(this.activaValue)
    const celdas = [
      // C23-11 · Con varias cajas abiertas la fila tiene que decir en cuál
      // entró; si no, la tabla mezcla las tres sin distinguirlas.
      ["px-6 py-3 text-sm font-bold text-cec-navy dark:text-cec-gold", fila.caja],
      ["px-6 py-3 text-sm font-mono text-cec-navy", fila.recepcion],
      ["px-6 py-3 text-sm font-mono text-gray-500", fila.tracking],
      ["px-6 py-3 text-sm text-gray-700", fila.cliente],
      ["px-6 py-3 text-sm text-gray-500", fila.tipo]
    ]
    celdas.forEach(([cls, texto]) => {
      const td = document.createElement("td")
      td.className = cls
      td.textContent = texto || "—"
      tr.appendChild(td)
    })
    this.filasTarget.prepend(tr)
  }

  _post(url, cuerpo) {
    const token = document.querySelector("meta[name='csrf-token']")?.content
    return fetch(url, {
      method: "POST",
      headers: { "Content-Type": "application/json", "Accept": "application/json", "X-CSRF-Token": token },
      body: JSON.stringify(cuerpo)
    }).then((r) => r.json())
  }
}

File: app/controllers/empaque_controller.rb
# C21-01 · El escaneo al empacar — el «pip pip pip».
#
# Yusef, mostrando la bodega en vivo por cámara mientras empacaban:
#
#   > "Ahí están empacando, mirá… **y aquí es donde hace falta, es el pip pip
#   >  pip**."
#
# Y la pregunta que abrió el módulo entero:
#
#   > "¿Qué otra forma puedo hacer para empezar a decir que **estos paquetes van
#   >  en esa caja**?"
#
# La `Fase 12` ya lo tenía dibujado desde la Conversación 5, con el detalle que
# importa: *"si el tipo de servicio no concuerda con el de la caja, pita"*, y un
# **botón de omitir** para no trabar la bodega cuando algo no cuadra.
#
# Es también el primer lugar del sistema que escribe el estado `empacado`. Vivía
# en el enum desde siempre y nadie lo asignaba: `EtiquetarController` lo dejó
# reservado con nombre y apellido (*"`empacado` queda reservado para el módulo
# de empaque, que todavía no existe"*).
class EmpaqueController < ApplicationController
  before_action :authorize_manifiestos
  before_action :set_manifiesto
  before_action :set_caja, only: %i[escanear]

  # C23-11 · Se empaca en **varias cajas a la vez**.
  #
  # Yusef, describiendo la bodega en temporada:
  #
  #   > "Pero aquí pues **debería de existir la múltiple** […] o sea, poder
  #   >  **seleccionar las tres cajas**."
  #   > "Es que **ellos arman tres cajas y empiezan a meter los paquetes en
  #   >  cualquier caja**."
  #
  # **Un paquete sigue yendo en UNA caja.** Jorge preguntó lo contrario —*"un
  # tracking puede tener tres cajas"*— y Yusef lo corrigió en el acto con ese
  # *"no"*: lo que pasa es que hay tres cajas **llenándose al mismo tiempo**, y
  # el paquete cae en la que tenga lugar. Así que no hay tabla de unión: lo que
  # cambia es la pantalla, no el modelo.
  #
  # El costo que él describió es el de ir a buscar la caja: *"tenés que ir a
  # buscar, o sea, hay que hacerlo bien el proceso"*. Elegir caja era un
  # `link_to` que **recargaba la pantalla**, y con eso el campo de escaneo
  # perdía el foco — con la pistola en la otra mano, eso es un clic por cada
  # cambio de caja.
  def show
    @cajas = @manifiesto.cajas.ordenadas
    @abiertas = cajas_abiertas
    @caja = caja_activa
    # La tabla muestra lo que hay en **todas las abiertas**, no en una: son las
    # que se están llenando a la vez, y verlas juntas es el punto.
    @paquetes = Paquete.where(caja_manifiesto_id: @abiertas.map(&:id))
                       .includes(:cliente, :tipo_envio, :caja_manifiesto)
                       .order(created_at: :desc)
  end

  # C23-11 · Abrir o cerrar una caja del set. *"Y para desempacarlo, lo volvemos
  # a marcar"* — es un interruptor, no dos botones.
  #
  # **Mínimo una**, que Jorge propuso y Yusef ratificó (*"máximo todas"*): sin
  # ninguna abierta el escaneo no tendría a dónde ir, y la pantalla quedaría
  # pidiendo un código que no puede guardar en ningún lado.
  def alternar
    caja = @manifiesto.cajas.find(params[:caja_id])
    abiertas = cajas_abiertas.map(&:id)

    if abiertas.include?(caja.id)
      if abiertas.size == 1
        return render json: { ok: false, mensaje: "Tiene que quedar al menos una caja abierta." }
      end
      abiertas -= [ caja.id ]
    else
      abiertas += [ caja.id ]
    end

    guardar_abiertas(abiertas)
    # Si se cerró la que estaba activa, el escaneo se muda a la primera que
    # quede abierta — nunca a una cerrada.
    activa = abiertas.include?(caja_activa_id) ? caja_activa_id : abiertas.first
    guardar_activa(activa)

    render json: { ok: true, abiertas: abiertas, activa: activa }
  end

  # Un escaneo. Contesta JSON porque el operario mira la pistola, no la
  # pantalla: lo que decide es el sonido, y la fila se agrega sin recargar.
  def escanear
    codigo = params[:codigo].to_s.strip
    paquete = buscar_paquete(codigo)

    return render json: { resultado: "no_encontrado", mensaje: "No se encontró ningún paquete con «#{codigo}»." } if paquete.nil?

    if ya_esta_en_otra_caja?(paquete)
      return render json: { resultado: "ya_empacado",
                            mensaje: "#{paquete.numero_recepcion_visible} ya está en la caja #{paquete.caja_manifiesto.letra}." }
    end

    unless tipo_permitido?(paquete)
      return render json: {
        resultado: "tipo_distinto",
        mensaje: "#{paquete.numero_recepcion_visible} es #{paquete.tipo_envio&.nombre || "sin tipo"}, " \
                 "y este manifiesto lleva #{@manifiesto.tipos_envio_nuestros}.",
        paquete_id: paquete.id
      }
    end

    empacar!(paquete)
    render json: { resultado: "ok", mensaje: "#{paquete.numero_recepcion_visible} entró a la caja #{@caja.letra}.",
                   fila: fila_de(paquete) }
  end

  # C21-01 · «Omitir»: mete el paquete igual, aunque el tipo no concuerde.
  # La `Fase 12` lo pidió con esas palabras — *"botón de omitir para no trabar
  # la operación cuando algo no cuadra"*. Queda en la bitácora del paquete, que
  # es donde se puede revisar después.
  def omitir
    @caja = @manifiesto.cajas.find(params[:caja_id])
    paquete = Paquete.find(params[:paquete_id])
    empacar!(paquete)
    render json: { resultado: "ok", mensaje: "#{paquete.numero_recepcion_visible} entró igual, omitiendo el aviso.",
                   fila: fila_de(paquete) }
  end

  private

  def authorize_manifiestos
    redirect_to root_path, alert: "No tienes permiso para acceder a esta seccion." unless can_access?(:manifiestos)
  end

  def set_manifiesto
    @manifiesto = Manifiesto.find(params[:manifiesto_id])
  end

  def set_caja
    @caja = @manifiesto.cajas.find(params[:caja_id])
  end

  # ── El set de cajas abiertas ─────────────────────────────────────────────
  #
  # Vive en la **sesión del servidor**, como el tipo de envío de `/etiquetar`:
  # es el estado de un turno de trabajo, no un dato del manifiesto. Recargar no
  # lo pierde, y no le pisa el set a la persona que empaca en la otra mesa.
  #
  # Por defecto están **todas abiertas**, que es «máximo todas» y es lo que
  # deja la pantalla usable sin configurar nada. Cerrar es cómo se achica.
  def cajas_abiertas
    todas = @manifiesto.cajas.ordenadas.to_a
    guardadas = session.dig(:empaque_abiertas, @manifiesto.id.to_s)
    return todas if guardadas.blank?

    # Se filtra contra las que existen: una caja borrada no puede quedar
    # abierta en la sesión de nadie. Y si no queda ninguna, vuelven todas.
    vivas = todas.select { |c| guardadas.include?(c.id) }
    vivas.presence || todas
  end

  def caja_activa
    abiertas = cajas_abiertas
    abiertas.find { |c| c.id == params[:caja_id].to_i } ||
      abiertas.find { |c| c.id == caja_activa_id } ||
      abiertas.first
  end

  def caja_activa_id
    session.dig(:empaque_activa, @manifiesto.id.to_s)
  end

  def guardar_abiertas(ids)
    session[:empaque_abiertas] = (session[:empaque_abiertas] || {}).merge(@manifiesto.id.to_s => ids)
  end

  def guardar_activa(id)
    session[:empaque_activa] = (session[:empaque_activa] || {}).merge(@manifiesto.id.to_s => id)
  end

  # El operario escanea **la etiqueta del paquete**, y ese código de barras es
  # el número de recepción con su sufijo de caja (`etiqueta_codigo_barras`), no
  # el tracking. Se prueban las dos cosas: el número y, si no, la escalera de
  # tracking que ya usa /etiquetar.
  def buscar_paquete(codigo)
    return nil if codigo.blank?

    base = codigo.split("-").first
    Paquete.find_by("UPPER(numero_recepcion) = ?", codigo.upcase) ||
      Paquete.find_by("UPPER(numero_recepcion) = ?", base.to_s.upcase) ||
      Paquete.buscar_escaneado(codigo).first
  end

  def ya_esta_en_otra_caja?(paquete)
    paquete.caja_manifiesto_id.present? && paquete.caja_manifiesto_id != @caja.id
  end

  # *"Si el tipo de servicio no concuerda con el de la caja, pita."* La caja
  # hereda los tipos del manifiesto: son los que el operario eligió al crearlo.
  def tipo_permitido?(paquete)
    @manifiesto.tipo_envio_ids.include?(paquete.tipo_envio_id)
  end

  def empacar!(paquete)
    paquete.update!(caja_manifiesto: @caja, manifiesto: @manifiesto,
                    estado: "empacado")
    @manifiesto.recalculate_totals!
  end

  def fila_de(paquete)
    {
      id: paquete.id,
      # C23-11 · Con varias cajas abiertas la fila tiene que decir **en cuál**
      # entró: si no, la tabla mezcla las tres y no se sabe qué se llenó.
      caja: "#{@caja.letra}#{@caja.numero_bulto}",
      recepcion: paquete.numero_recepcion_visible,
      tracking: paquete.tracking,
      cliente: paquete.cliente&.nombre_completo,
      tipo: paquete.tipo_envio&.nombre
    }
  end
end
